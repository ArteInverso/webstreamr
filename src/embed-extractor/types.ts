import { StreamWithMeta } from '../types';

export interface EmbedExtractor {
  readonly id: string;

  readonly label: string;

  readonly extract: (embedUrl: string, language: string) => Promise<StreamWithMeta>;
}
