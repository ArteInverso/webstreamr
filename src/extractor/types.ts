import { Context, CountryCode, UrlResult } from '../types';

export interface Extractor {
  readonly id: string;

  readonly label: string;

  readonly ttl: number;

  readonly supports: (ctx: Context, url: URL) => boolean;

  readonly normalize: (url: URL) => URL;

  readonly extract: (ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined) => Promise<UrlResult[]>;
}
