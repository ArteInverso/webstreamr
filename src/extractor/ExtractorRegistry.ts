import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Extractor } from './types';
import { Context, UrlResult } from '../types';
import { Fetcher } from '../utils';
import { DoodStream } from './DoodStream';
import { Dropload } from './Dropload';
import { SuperVideo } from './SuperVideo';
import { ExternalUrl } from './ExternalUrl';
import { NotFoundError } from '../error';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult | undefined>;

  constructor(logger: winston.Logger, fetcher: Fetcher) {
    this.logger = logger;
    this.extractors = [
      new DoodStream(fetcher),
      new Dropload(fetcher),
      new SuperVideo(fetcher),
      new ExternalUrl(fetcher), // fallback extractor which must come last
    ];
    this.urlResultCache = new TTLCache({ max: 1024 });
  }

  readonly handle = async (ctx: Context, url: URL, countryCode: string): Promise<UrlResult | undefined> => {
    let urlResult = this.urlResultCache.get(url.href);
    if (this.urlResultCache.has(url.href)) {
      return urlResult;
    }

    const extractor = this.extractors.find(extractor => extractor.supports(url)) as Extractor;
    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    try {
      urlResult = await extractor.extract(ctx, url, countryCode);
    } catch (error) {
      if (error instanceof NotFoundError) {
        this.urlResultCache.set(url.href, urlResult, { ttl: extractor.ttl });

        return undefined;
      }

      this.logger.warn(`${extractor.id} error: ${error}`, ctx);
      return undefined;
    }

    this.urlResultCache.set(url.href, urlResult, { ttl: extractor.ttl });

    return urlResult;
  };
}
