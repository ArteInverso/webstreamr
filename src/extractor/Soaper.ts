import { Extractor } from './types';
import { Fetcher } from '../utils';
import { Context, Meta, UrlResult } from '../types';

interface SoaperInfoResponsePartial {
  val: string;
  val_bak: string;
}

export class Soaper implements Extractor {
  readonly id = 'soaper';

  readonly label = 'Soaper';

  readonly ttl = 900000; // 15m

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (_ctx: Context, url: URL): boolean => null !== url.host.match(/soaper/) && null !== url.pathname.match(/^\/(episode|movie)_/);

  readonly normalize = (url: URL): URL => url;

  readonly extract = async (ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> => {
    const movieOrEpisodeId = (url.pathname.match(/\/\w+_(\w+)/) as string[])[1] as string;

    const form = new URLSearchParams();
    form.append('pass', movieOrEpisodeId);
    form.append('param', '');
    form.append('extra', '');
    form.append('e2', '');
    form.append('server', '0');

    const response = await this.fetcher.textPost(
      ctx,
      new URL(url.pathname.includes('episode') ? '/home/index/GetEInfoAjax' : '/home/index/GetMInfoAjax', url.origin),
      form.toString(),
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      },
    );
    const jsonResponse = JSON.parse(response) as SoaperInfoResponsePartial;

    const m3u8Url = new URL(jsonResponse['val'], url.origin);
    const m3u8Data = await this.fetcher.text(ctx, m3u8Url);
    const height = m3u8Data.match(/\d+x(\d+)|(\d+)p/) as string[];

    return [
      {
        url: new URL(jsonResponse['val'], url.origin),
        label: this.label,
        sourceId: `${this.id}_${meta.countryCode.toLowerCase()}`,
        ttl: this.ttl,
        meta: {
          ...meta,
          ...(height && { height: parseInt(height[1] ?? height[2] as string) }),
        },
      },
    ];
  };
}
