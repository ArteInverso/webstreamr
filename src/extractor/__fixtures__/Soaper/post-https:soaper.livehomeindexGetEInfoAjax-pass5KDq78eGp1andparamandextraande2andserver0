{"key":true,"val":"\/home\/index\/TVM3U8?key=RnX8mnL4l5hEz38dxBAbFjV27jbe3Yh13LqxV48ViM7j60OLjmh3w5PJ8nQPh2brz0eEAAt50Zvv7d89hpM3OLWQ6nsjb4n0l3ApSna13LB3EbUxbbyJ70KwuVjaooMP5mIE4MbjyvqJHjRq3amYA5SPV8NRl8wmivl8YyrlMmFx6VomOAp2C5WR46w10v.m3u8","vtt":"https:\/\/ttt5.randchange.top\/\/hls\/2011\/black_mirror\/Season4\/2\/vtt\/\/thumb.vtt","val_bak":"\/home\/index\/TVM3U8?key=RnX8mnL4l5hEz38dxBAbFjV27jbe3Yh13LqxV48ViM7j60OLjmh3w5PJ8nQPh2brz0eEAAt50Zvv7d89hpM3OLWQ6nsjb4n0l3ApSna13LB3EbUxbbyJ70KwuVjaooMP5mIE4MbjyvqJHjRq3amYA5SPV8NRl8wmivl8YyrlMmFx6VomOAp2C5WR46w10v.m3u8","pos":0,"type":"m3u8","subs":[{"path":"\/home\/index\/getsrt?key=21ZbK1AB7YHPvy2d365Qf7A6d7Ox9jfbmPWyprvpf10rL2xor3Cndq3BJva3cP3Ovd0LEECvdPLL1r03cLX7m9BWybFqrw67eP2mH2q0E4ZE71UbrrWQaMo1tJyY226R5EhRalNQPp1dslnMX7bwpNs1LJqWbJV2TWK9EZdK3jU87lwBMprYI7R84zBEpKhXqmEYa5xQF5PbmA6KwBS86nXW4X1oh6K7y5X2O4HX8qPV76opTyB73AJPV4T215N5bRv2SrMZjKwb43HXmmJqRexPs2zZ6n2vAdfREPpP2zOyCOLnz4YqoVHyQWMyXNJMCO7q9yQZMVTQ4aMwzZ8YIAxBy7a7YbcxdA2M4vNLSjyqNA9E4Ls45jV04zvnc5AO4nNLvKsn6v9pK6jmSVdBlq3Lz4SPdKYJW.srt","name":"en"},{"path":"\/home\/index\/getsrt?key=nv0KMvQzYRC8wZz5KRELirnNMraXJRcn4VqYd02dUxW7y5K379iZ2d56W7r5c7Me612055FQ8m99weWAIM0VlyE89zU0QwN81EAVIMeWBE8Bx6Fann2BRWXzf78mwwny6oir8x5BMVJ6HeaYn3EBjwFXVLox8L36sK7x9o67mvcE15vxZzb9SqmwQ4PKZnCpyA8ojR15ilyaB6vxrRcZzvPYLPbqSJOYVvqAK4T94Z0nYNvAH1QJ8LYnvzSlorwr4KylFb5aAp3LQYiBRRn417VpSdY76Mdl9WuylEzEKjPYCMXZ0V9qexh0xBW0N3wWSK9p51wJVviA4yN3w0m2Iq01r2d2WQInYajrJbovIpoB4EK38eip1Je6pKvxubN53JZwE7h9YzyjOYdAULxabZbYVRCnbAnr5Mvqip1q5WO.srt","name":"en:hi"},{"path":"\/home\/index\/getsrt?key=Lre35r89YAh4O1KRQe0duLdBzLNyRaf8zPmE6qo6S1Mrj60yrmCvAoKOqY2KT5Kxp09ennUz20YYPQ1yu3wROMp4LyTYRA4Xyq0Ziw4rAnJA3ahnddr1mKyOF6Ja77xLmyuJ5ZMpoxrnfBlonvEWrzS1YXqMwXd6T7LbVjqLRwu0rVoWvljzhpZ5q6bvPdTALJaV0rbQHNxRBr316zf1pKN75Nq2i7MoEdqjZRi024moe1yNcMdmrq4xjptnA3P3O7enT1xQvYzyBphynnBJeOAluBvpONBbz3IQNR4RjZ6Wu9QqV1dKoyfeNVjeoqnjCdAyMbRLnrUwvYX14eKMtj0rZ2b2APIMjAd2OJrotXVAJ0LNE1Hpwn7Lp0O6ulYVn6x0bQtbe6d5MPM7CB7QqwKqWmIjz2X2lnQKFleErlJB6oSbNWv908YqinMZ1JM4KmtO79pOXZR6FYePVNNoyNHLV4Q0o93EI5nrP1KvLNFK0Qj66qOOh7Aw46LP5Ps1nQoAyELoFme4MzM53dUzy9.srt","name":"Encoded By nItRo And Uploaded By XpoZ.srt"}],"prev_epi_title":"[S4E1] USS Callister","prev_epi_url":"\/episode_4vGPvjBD31.html?ap=1","next_epi_title":"[S4E3] Crocodile","next_epi_url":"\/episode_PnG6nAnG7v.html?ap=1"}