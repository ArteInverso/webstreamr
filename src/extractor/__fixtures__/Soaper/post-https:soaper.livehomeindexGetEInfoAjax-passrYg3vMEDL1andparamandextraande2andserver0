{"key":true,"val":"\/home\/index\/TVM3U8?key=LrezaYpLOxT4O61m8e3mSLdBzLNyRaf8zPmERxvwh1Mrj6zbVEUvAzRv2KZLu5PVvRKqwbcz6qmb9vbOi3wQZ4ae8RsZ9N3Jm9Yatw4yE55bBp.m3u8","vtt":"https:\/\/ttt5.randchange.top\/\/hls\/2023\/last_of_us\/Season2\/3\/vtt\/\/thumb.vtt","val_bak":"\/home\/index\/TVM3U8?key=LrezaYpLOxT4O61m8e3mSLdBzLNyRaf8zPmERxvwh1Mrj6zbVEUvAzRv2KZLu5PVvRKqwbcz6qmb9vbOi3wQZ4ae8RsZ9N3Jm9Yatw4yE55bBp.m3u8","pos":0,"type":"m3u8","subs":null,"prev_epi_title":"[S2E2] Through the Valley","prev_epi_url":"\/episode_dZkrY8nG6z.html?ap=1","next_epi_title":"[S2E4] Day One","next_epi_url":"\/episode_OMkKZ9WGLq.html?ap=1"}