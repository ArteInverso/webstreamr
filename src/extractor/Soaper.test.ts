import winston from 'winston';
import { Fetcher } from '../utils';
import { Context, CountryCode } from '../types';
import { ExtractorRegistry } from './ExtractorRegistry';
import { Soaper } from './Soaper';

jest.mock('../utils/Fetcher');

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
// @ts-expect-error No constructor args needed
const fetcher = new Fetcher();
const extractorRegistry = new ExtractorRegistry(logger, [new Soaper(fetcher)]);

const ctx: Context = { id: 'id', ip: '127.0.0.1', config: {} };

describe('Soaper', () => {
  test('Full Metal Jacket', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://soaper.live/movie_d8kdeypDY9.html'), CountryCode.en, 'Full Metal Jacket (1987)')).toMatchSnapshot();
  });

  test('Black Mirror', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://soaper.live/episode_5KDq78eGp1.html'), CountryCode.en, 'Black Mirror 4x2')).toMatchSnapshot();
  });
});
