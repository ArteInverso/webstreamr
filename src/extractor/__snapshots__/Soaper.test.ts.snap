// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Soaper Black Mirror 1`] = `
[
  {
    "label": "Soaper",
    "meta": {
      "countryCode": "en",
      "height": 720,
      "title": "Black Mirror 4x2",
    },
    "sourceId": "soaper_en",
    "ttl": 900000,
    "url": "https://soaper.live/home/index/TVM3U8?key=zoeyOo54b7s8NAPnOW0div4OdvmbYxIzpE2RnmAoSXqMAm7eZKs79V4lxEnZsvQ6qYNrOEiqKO4dAwdBiYx3KmwyVRc9oPzLWoOYcLQVArr2j0.m3u8",
  },
]
`;

exports[`Soaper Full Metal Jacket 1`] = `
[
  {
    "label": "Soaper",
    "meta": {
      "countryCode": "en",
      "height": 1080,
      "title": "Full Metal Jacket (1987)",
    },
    "sourceId": "soaper_en",
    "ttl": 900000,
    "url": "https://soaper.live/home/index/M3U8?key=VnlzaOr0NYimoY5BLezvsJ8zVLzvdAUVW6XKL711TqadxBBy7AheAYoXRl28s78dmP5WOQUVdxX9RWdOHNRanKJaQ7F5mwRpaWN5fPdMm8M.m3u8",
  },
]
`;
